core/src/main/scala/com.mediative.amadou/Stage.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 Mediative
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mediative.amadou
+
+import org.apache.spark.sql._
+import scala.util.{ Failure, Try }
+
+sealed trait Stage[-I, +T] { self =>
+  def name: String
+
+  def map[U](f: T => U): Stage[I, U] = new Stage[I, U] {
```

Comment:
So, every map over a stage will create a new anonymous class. I would recommend putting a mapper implementation class in the `Stage` companion object to avoid this, e.g. `object Stage { private class Mapper[-I, T, +U](f: T => U) extends Stage[I, U] { ... } }`.

Then the `Stage#map` operation can be just `new Mapper[I, T, U](f)`.